### PR TITLE
Get the YAML and logs for a single failed Pod, if Pods are not marked as ready/complete in time.

### PR DESCRIPTION
**Pull request by Harwayne**
**What this PR does, why we need it**:
Get the YAML and logs for a single failed Pod, if Pods are not marked as ready/complete in time.

**User-visible changes in this PR**:
If wait_until_pods_running fails, then it will log the YAML and logs for a single one of the failed pods.

### CODE REVIEW

**Comment by adrcunha** [comment]
@Harwayne is intended that this PR is marked as draft?

**Comment by chaodaiG** [comment]
This would be pretty helpful!
/approve

**Comment by Harwayne** [comment]
> @Harwayne is intended that this PR is marked as draft?

I marked it as a draft as I haven't seen it work yet.

I created it because some automated tests were failing just saying 'not all pods in knative-eventing got to RUNNING' (or something to that effect). I wrote this change and wanted to see the real error, but since then the automated tests have failed for other reasons (couldn't create the cluster).

If you are confident in the change, I can submit as-is. I am not confident in my bash writing skills, so wanted to see it work first.

**Comment by adrcunha** [comment]
> I marked it as a draft as I haven't seen it work yet.

Force an error on line 146 by changing the status variable, and run e2e-tests.sh locally.

File: scripts/library.sh [hunk 1]
```diff
@@ -132,30 +132,47 @@ function wait_until_object_does_not_exist() {
...
       while read pod ; do
         local status=(`echo -n ${pod} | cut -f2 -d' ' | tr '/' ' '`)
+        local pod_name=$(echo -n ${pod} | cut -f1 -d' ')
         # All containers must be ready
-        [[ -z ${status[0]} ]] && all_ready=0 && break
-        [[ -z ${status[1]} ]] && all_ready=0 && break
-        [[ ${status[0]} -lt 1 ]] && all_ready=0 && break
-        [[ ${status[1]} -lt 1 ]] && all_ready=0 && break
-        [[ ${status[0]} -ne ${status[1]} ]] && all_ready=0 && break
+        [[ -z ${status[0]} ]] && all_ready=0 && failed_pod="${pod_name}" && break
```
Comment (adrcunha) [inline scripts/library.sh]: You can simplify this series of checks by setting failed_pod in line 146 only, then resetting it to empty string in line 153.

File: scripts/library.sh [hunk 2]
```diff
@@ -132,30 +132,47 @@ function wait_until_object_does_not_exist() {
...
+        [[ ${status[1]} -lt 1 ]] && all_ready=0 && failed_pod="${pod_name}" && break
+        [[ ${status[0]} -ne ${status[1]} ]] && all_ready=0 && failed_pod="${pod_name}" && break
       done <<< "$(echo "${pods}" | grep -v Completed)"
       if (( all_ready )); then
         echo -e "\nAll pods are up:\n${pods}"
         return 0
       fi
+    elif [[ ${not_running} -ne 0 ]]; then
+      # At least one Pod is not running, just save the first one's name as the failed_pod.
+      while read pod ; do
```
Comment (adrcunha) [inline scripts/library.sh]: You can replace this `while` with just

```
failed_pod="$(echo "${not_running_pods}" | head -n 1 | cut -f1 -d' ')"
```

File: scripts/library.sh [hunk 3]
```diff
@@ -132,30 +132,47 @@ function wait_until_object_does_not_exist() {
...
+        failed_pod="${pod_name}"
+        break
+      done <<< "$(echo "${not_running_pods}")"
     fi
     echo -n "."
     sleep 2
   done
   echo -e "\n\nERROR: timeout waiting for pods to come up\n${pods}"
+  if [[ -n "${failed_pod}" ]]; then
+    echo -e "\n\nFailed pod oyaml - ${failed_pod}\n"
```
Comment (adrcunha) [inline scripts/library.sh]: "Failed pod oyaml" sounds like a typo to me. :) Maybe something like "Failed pod (data in YAML format)"?

**Comment by Harwayne** [comment]
Manually tested in the knative/eventing repo.

**Pull request merged**